Add support for primitive types `char` and 128-bit integers

ssyram opened the pull request:
The type `char` is now simply uint32_t. And the types `i128` and `u128` are now supported. Now the missing parts are the binary operations. Will add the support subsequently.

Now, the 128-bit literals are supported as built-in functions from bits. E.g., `0x1234567891011121314151617181920u128` is represented as: `Eurydice_u128_from_bits(0x123456789101112, 0x1314151617181920)`, which has the simple implementation of `(Eurydice_uint128_t) high << 64 | low`. Note that this division is needed anyway, even when Krml supports 128-bit length -- as C compilers will complain if we have literals larger than `uint64_t`.

I don't know yet where to store the primitive implementation, to pass the tests, I placed the implementations now within the `test/main.c`, which accompanies the compilation for the tests.

Another point to note is the support of `SwitchInt` for the 128 bits integers -- as Krml does not support this, the expression is expressed as normal if-then-else expressions.

Also, I'm curious about where to place these primitive implementations (e.g., `Eurydice_u128_add`) -- I can now only declare `extern` in the generated codes for the built-in functions. I need a place to provide implementations for these built-in functions.

ssyram commented:
Full support for the 128-bit integers are now added, now the operations are translated as built-in functions for these types. Please consider merging.

msprotz commented on lib/AstOfLlbc.ml:
```diff
@@ -794,27 +836,108 @@ let op_of_binop (op : C.binop) : Krml.Constant.op =
+          else uint128_op_of_op op
+        in
+        let ret_t, _ = Krml.Helpers.flatten_arrow op.typ in
+        op, ret_t
```
this seems like the right thing to do

msprotz commented on lib/AstOfLlbc.ml:
```diff
@@ -794,27 +836,108 @@ let op_of_binop (op : C.binop) : Krml.Constant.op =
+  (* Modify here *)
+  let rest =
+    match op.node with
+    | EOp (BShiftL,_) | EOp (BShiftR,_) -> modify_rest rest
```
is this not already handled in lib/PreCleanup.ml with `adjust_shift`? could you extend this phase? I don't have a strong opinion either way, let's just settle on one single way to do this

ssyram replied:
Yes, I found it more robust for at least i128/u128 to handle the case here (I do not trust the modification of AST in the Cleanup phase -- I don't want to post additional assumptions on these Cleanups, and I want to handle the case at its source, which would definitely do the right thing). And it would be more natural to handle also usual `Shift` here, so for robustness, I also add this.

ssyram replied:
OK

msprotz commented on lib/Builtin.ml:
```diff
@@ -42,6 +47,508 @@ let expr_of_builtin { name; typ; cg_args; _ } =
+      cg_args = [];
+      arg_names = [ "lhs"; "rhs" ];
+    }
+
```
ok there's got to be a way to generate these using a big `List.map`, no? like `List.map (mk_binop uint128_t) [ "add"; "sub"; "mul"; "div"; ... ] @ List.map (mk_binop int128_t) [ "add"; "sub"; ... ]`

ssyram replied:
Thanks for the advice! Will do.

msprotz reviewed (commented):
Overall, this is a great PR -- I'm excited for this extended support! Some comments, (which should also address some of your questions on zulip).
- The deciding criterion as to whether something goes as an integer type in https://github.com/FStarLang/karamel/blob/master/lib/Constant.ml#L10 is whether it is supported by primitive C operators. This means that Float and Double should definitely go in there (feel free to pick better names, like Float32 and Float64). However, I would not add U128 and I128 on the basis that U128 is already library-defined (see my other comment inline), and I would not add I128 either for that reason.
- Where to place the implementations: one thing I'd like to do is split eurydice_glue.h into several headers, like `eurydice/vec.h`, `eurydice/slice.h`, `eurydice/array.h`, `eurydice/box.h`, etc. and then you could have `eurydice/int128.h` provided you mark those definitions as `static inline` (possibly setting Krml.Options.static_header to contain the definitions from `Eurydice_UInt128`

Overall this PR seems pretty ready to me! At a high level:
- let's decide on whether uint128 ought to be a primitive type or not
- let's do something more concise for Builtin.ml, it'll also be easier to maintain
- let's figure out how to make these uint128 definitions go into a "static header" (so that they can appear in eurydice_glue.h)

And then we'll be good to go? Let me know if there's anything else we need to discuss before merging. Cheers and thanks again for a great PR

ssyram commented on include/eurydice_glue.h:
```diff
@@ -14,6 +14,11 @@
+
+typedef __int128_t Eurydice_int128_t;
+typedef __uint128_t Eurydice_uint128_t;
+
```
Yes, I think this would be a good idea -- instead of primitive, I think using the current version but with a different "eurydice_glue.h" implementation by adding some `#if` would be a better idea. Then, I think we might also support `f16` / `f128` in a near way, too? `f32` / `f64` should definitely be primitive.

In total, I would like to do the following in "eurydice_glue.h":
#if ...
Eurydice_int128_t int128_from_bits(uint64_t high, uint64_t low) {
    return (Eurydice_int128_t) { .high = high, .low = low };
}
#...
But what should be the branch case to test?

Btw, I'm thinking about using `f16_from_str("1.5")` as the initialization of `f16`. Then, there might be some primitive implementation of these functions.

ssyram commented:
Thanks for the detailed comment! I have the following views:
1. it would definitely be a good idea to split the currently complex "eurydice_glue.h", but it involves additionally a functionality of judging which file to import? This might go over the current work PR, so may I simply place the `static inline` stuff in "eurydice_glue.h" for now with an additional comment?
2. Like in the inline comment, I'm also thinking about supporting `f16` and `f128`, would it be a good idea to write the literals as a function from the literal string?

ssyram commented on include/eurydice_glue.h:
```diff
@@ -513,3 +513,414 @@ typedef struct {
+  return result;
+}
+static inline Eurydice_int128_t Eurydice_i128_neg(Eurydice_int128_t x) {
+  return *(Eurydice_int128_t*)&Eurydice_u128_neg(*(Eurydice_uint128_t*)&x);
```
Their internal representations are the same, so here we just do a type conversion.

msprotz replied:
it is not, see the links in the other comment

ssyram commented:
Thanks! Will make the adjustments soon!

ssyram commented on out/test-more_primitive_types/internal/Eurydice.h:
```diff
@@ -0,0 +1,97 @@
+extern "C" {
+#endif
+
+extern Eurydice_int128_t Eurydice_i128_add(Eurydice_int128_t lhs, Eurydice_int128_t rhs);
```
OK, I will try to find the places to accommodate the changes.

ssyram commented on lib/AstOfLlbc.ml:
```diff
@@ -613,16 +615,56 @@ let expression_of_var_id (env : env) (v : C.local_id) : K.expr =
-  K.(with_type (TInt w) (EConstant (constant_of_scalar_value sv)))
+(** Assume here the maximum length is 128-bit -- throw away any bits outside the initial 128 bits
+    Returns the **expr** pair (high64bits, low64bits) *)
+let split_128bit (value : Z.t) =
```
As this is simply a helper function, I want it to be more robust -- as the input type is Z.t, it is possible that the length will exceed 128 bits. I want to make this helper function robust enough, instead of throwing errors.

ssyram commented:
Revision made:
- 128-bit related built-ins are moved to "include/eurydice/int128.h";
- changed the comments of "split_128bit";
- Used "Krml.Helpers.assert_elid";
- Removed adjust_shift from PreCleanup;
- Added the cleanup of removing the built-in functions as externals.

msprotz reviewed (commented):
Ok we're nearly there.
- Please REMOVE the old shift adjustment phase in `PreCleanup.ml`. This is going to be very confusing if two pieces of code are trying to do the same thing.
- We already have a way to eliminate prototypes from the generated C code, so `remove_builtin_funcs` needs to go. It works, but it duplicates logic that exists elsewhere.

What I suggest for the latter point is:
- in Builtin.ml, line 57: replace `["Eurydice"]` with `["Eurydice";"Int128"]`
- in `bin/main.ml`, line 90, add `Bundle.Prefix ["Eurydice"; "Int128"]` to the list

This will eliminate confusion (why have two different ways of doing the same thing?), and will also make code easier to maintain in the long run.

Thank you!

msprotz commented on include/eurydice_glue.h:
```diff
@@ -513,3 +513,414 @@ typedef struct {
+}
+static inline
+Eurydice_int128_t Eurydice_i128_band(Eurydice_int128_t lhs, Eurydice_int128_t rhs) {
+  return *(Eurydice_int128_t*)&Eurydice_u128_band(*(Eurydice_uint128_t*)&lhs, *(Eurydice_uint128_t*)&rhs);
```
I know that they have the same representation, but it's illegal in C and will potentially cause incorrect code to be generated

see https://blog.regehr.org/archives/1520 and https://gist.github.com/shafik/848ae25ee209f698763cffee272a58f8 for some reading

so unfortunately, you just can't do that

msprotz commented on include/eurydice_glue.h:
```diff
@@ -513,3 +513,414 @@ typedef struct {
+  uint64_t p3 = a_hi * b_hi;
+  uint64_t carry = ((p0 >> 32) + (uint32_t)p1 + (uint32_t)p2) >> 32;
+  *lo = p0 + ((p1 + p2) << 32);
+  *hi = p3 + (p1 >> 32) + (p2 >> 32) + carry;
```
I am not integrating AI-generated code in a project that is about verified software, unless it comes with a proof of correctness -- right now, I have no way to convince myself that this piece of code is always computing the correct result, and testing all possible combinations is not feasible

ssyram replied:
Maybe we can fix this implementation later? As this is only a matter of fixing the C codes, which is totally decoupled from the OCaml codes.

ssyram commented on lib/Cleanup3.ml:
```diff
@@ -171,3 +171,12 @@ let also_skip_prefix_for_external_types (scope_env, _) =
           KPrint.bprintf "Warning! The skip_prefix options generate name conflicts\n"
   end
+
+let remove_builtin_funcs files =
```
OK, I did not find this. I will try.

msprotz commented:
Separately, @Nadrieril and I are trying to extract https://github.com/rust-lang/compiler-builtins/tree/master/compiler-builtins/src/int/ which would give us the same implementation of 128-bit arithmetic as the Rust compiler itself, and then we could make your int128 header use that in case there is no built-in 128-bit type in the current C toolchain

ssyram commented:
How can I make `uint128` optional?

msprotz commented:
> Header name fixed. How can I make `uint128` optional? The error seems to lie in that `eurydice/int128.h` does not exist. Is this a problem with the auto check procedure?

This is how I did it: https://github.com/AeneasVerif/eurydice/blob/main/test/more_primitive_types.yaml -- I made sure eurydice_glue.h does *not* include eurydice/int128.h, but instead, the user can pass a "config" file that indicates that some C file should contain and extra #include directive.

Let me know if you think that doesn't work but I think it's better because it makes eurydice_glue.h smaller, and paves the way for a more modular separation of all the stuff in eurydice_glue.h

ssyram commented:
OK, so now int128 and u128 are all supported?

msprotz commented:
Yes all of your commits on this PR up to [d7cce62](https://github.com/AeneasVerif/eurydice/pull/195/commits/d7cce62103c91d85671f8e1e251a0d0c1557e93f) have been merged

ssyram commented:
That's great, thanks!
